File: src/agent-app-node/src/searchClient.js
/**
 * Azure AI Search client for vector/hybrid search.
 */
import { SearchClient } from '@azure/search-documents';
import { DefaultAzureCredential } from '@azure/identity';
import { getTracer } from './telemetry.js';
import { config } from './config.js';

const tracer = getTracer('macu-search-client');

/**
 * Search the MACU knowledge base using vector search.
 */
export class KnowledgeBaseSearch {
  constructor() {
    const credential = new DefaultAzureCredential();
    this.client = new SearchClient(
      config.searchEndpoint,
      config.searchIndexName,
      credential
    );
  }

  /**
   * Hybrid search: keyword + vector similarity.
   * @param {string} query - Search query text
   * @param {number[]} embedding - Query embedding vector
   * @param {number} topK - Number of results to return
   * @returns {Promise<Array<{title: string, content: string, source: string, score: number}>>}
   */
  async search(query, embedding, topK = 5) {
    const span = tracer.startSpan('knowledge-base.search');

    try {
      span.setAttribute('query', query);
      span.setAttribute('top_k', topK);
      span.setAttribute('index', config.searchIndexName);

      const searchResults = await this.client.search(query, {
        vectorSearchOptions: {
          queries: [
            {
              kind: 'vector',
              vector: embedding,
              kNearestNeighborsCount: topK,
              fields: ['content_vector'],
            },
          ],
        },
        select: ['title', 'content', 'source'],
        top: topK,
      });

      const documents = [];
      for await (const result of searchResults.results) {
        documents.push({
          title: result.document.title || '',
          content: result.document.content || '',
          source: result.document.source || '',
          score: result.score || 0,
        });
      }

      span.setAttribute('results.count', documents.length);
      return documents;
    } catch (error) {
      span.recordException(error);
      throw error;
    } finally {
      span.end();
    }
  }
}

File: src/agent-app/search_client.py
"""
Azure AI Search client for vector/hybrid search.
"""
from azure.search.documents import SearchClient
from azure.search.documents.models import VectorizedQuery
from azure.identity import DefaultAzureCredential
from opentelemetry import trace
from config import Config

tracer = trace.get_tracer("macu-search-client")


class KnowledgeBaseSearch:
    """Search the MACU knowledge base using vector search."""

    def __init__(self):
        credential = DefaultAzureCredential()
        self.client = SearchClient(
            endpoint=Config.SEARCH_ENDPOINT,
            index_name=Config.SEARCH_INDEX_NAME,
            credential=credential,
        )

    def search(self, query: str, embedding: list[float], top_k: int = 5) -> list[dict]:
        """Hybrid search: keyword + vector similarity."""
        with tracer.start_as_current_span("knowledge-base.search") as span:
            span.set_attribute("query", query)
            span.set_attribute("top_k", top_k)
            span.set_attribute("index", Config.SEARCH_INDEX_NAME)

            results = self.client.search(
                search_text=query,
                vector_queries=[
                    VectorizedQuery(
                        vector=embedding,
                        k_nearest_neighbors=top_k,
                        fields="content_vector",
                    )
                ],
                select=["title", "content", "source"],
                top=top_k,
            )

            documents = []
            for result in results:
                documents.append({
                    "title": result.get("title", ""),
                    "content": result.get("content", ""),
                    "source": result.get("source", ""),
                    "score": result.get("@search.score", 0),
                })

            span.set_attribute("results.count", len(documents))
            return documents